### DISCUSSION TITLE
ChatGPT Voices

### DISCUSSION THREAD

**Discussion by danfrist** [jing332/tts-server-android#143] [General]
I love this app, and I use it every day. I am not visually impaired, either. I just use it for productivity and reading retention.

Question: Would it be possible to add ChatGPT voices?

**Comment by danfrist** [comment]
Here is an example of a successful call to the OpenAI API:

Request URL: https://api.openai.com/v1/audio/speech
Request Method: POST
Payload: {"model":"tts-1","input":"The text to speak.","voice":"nova","response_format":"opus"}

Headers: 
Authorization: Bearer API-KEY
Content-Type: application/json

I tried a couple things with Custom TTS in the app, but I get a message that the auth header is not being sent, even though I've got it in the Http Request header field.

**Comment by jing332** [comment]
Custom TTS has been deprecated. Use plugin instead.

``` javascript

// EDIT: let apiKey = ttsrv.userVars["key"]
let apiKey = ttsrv.userVars["apiKey"]


let PluginJS = {
    'name': 'ChatGPT',
    'id': 'speech.openai.com',
    'author': 'TTS Server',
    'description': '',
    'version': 1,
    "vars": {
        apiKey: {label: "API-KEY", hint: "OpenAI API-KEY"},
    },


    'getAudio': function (text, locale, voice, speed, volume, pitch) {
        let reqHeaders = {
            'Content-Type': 'application/json',
            'Authorization': 'Bearer ' + apiKey
        }

        let body = {"model":"tts-1", "input":text, "voice":voice, "response_format":"opus"}
        let str = JSON.stringify(body)
        let resp = ttsrv.httpPost('https://api.openai.com/v1/audio/speech', str, reqHeaders)

        if (resp.isSuccessful()) {
            return resp.body().byteStream()
        } else {
            throw "FAILED: status=" + resp.code()
        }
    }
}

let EditorJS = {
    'getAudioSampleRate': function (locale, voice) {
         let audio = PluginJS.getAudio('test', locale, voice, 50, 50, 50)
         return ttsrv.getAudioSampleRate(audio)
//        return 22050
    },


    'onLoadData': function () {

    },

    'getLocales': function () {
        return ['en-US']
    },

    'getVoices': function (locale) {
      return {
        'nova': 'Nova',
      }
    },

    'onLoadUI': function (ctx, linearLayout) {

    }

}

```

**Comment by jing332** [comment]
Use devlopment version from Github actions

**Comment by danfrist** [comment]
Thank you for your reply!

Unfortunately, I received the following errors:

com.script.ScriptException: org.mozilla.javascript.JavaScriptException:
FAILED: status=401 (<Unknown source>#27) in <Unknown source> at line number
27
 at
com.script.javascript.RhinoScriptEngine.invoke(RhinoScriptEngine.java:107)
 at
com.script.javascript.RhinoScriptEngine.invokeMethod(RhinoScriptEngine.java:3)
 at ra.c.d(TtsPluginUiEngine.kt:35)
 at u9.y0$a$c.invokeSuspend(PluginTtsUI.kt:34)
 at ff.a.resumeWith(ContinuationImpl.kt:9)
 at eg.p0.run(DispatchedTask.kt:111)
 at jg.j$a.run(LimitedDispatcher.kt:4)
 at lg.j.run(Tasks.kt:3)
 at lg.a$a.run(CoroutineScheduler.kt:96)
Caused by: org.mozilla.javascript.JavaScriptException: FAILED: status=401
(<Unknown source>#27)
 at org.mozilla.javascript.Interpreter.interpretLoop(Interpreter.java:77)
 at org.mozilla.javascript.Interpreter.interpret(Interpreter.java:57)
 at
org.mozilla.javascript.InterpretedFunction.call(InterpretedFunction.java:21)
 at org.mozilla.javascript.ContextFactory.doTopCall(ContextFactory.java:1)
 at
com.script.javascript.RhinoScriptEngine$1.superDoTopCall(RhinoScriptEngine.java:1)
 at
com.script.javascript.RhinoScriptEngine$1.doTopCall(RhinoScriptEngine.java:40)
 at org.mozilla.javascript.ScriptRuntime.doTopCall(ScriptRuntime.java:8)
 at
org.mozilla.javascript.InterpretedFunction.call(InterpretedFunction.java:16)
 at
com.script.javascript.RhinoScriptEngine.invoke(RhinoScriptEngine.java:60)
 ... 8 more


https://bin.kv2.dev/~65b7b2205de7ed6bfdbca56f

**Reply by mrvson** [reply]
I also got the same error

**Reply by mrvson** [reply]
My gpt account is no longer plus so it has limited api calls

**Comment by danfrist** [comment]
If you need a ChatGPT key for testing, please provide a private way to contact you, and I will send one.

**Comment by jing332** [comment]
401 unauthorized


Have you set up the plugin variables?




---Original---
From: ***@***.***&gt;
Date: Mon, Jan 29, 2024 22:12 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [jing332/tts-server-android] ChatGPT Voices (Discussion #143)




 Thank you for your reply! 
 
 Unfortunately, I received the following errors: 
 
 com.script.ScriptException: org.mozilla.javascript.JavaScriptException: 
 FAILED: status=401 (<Unknown source&gt;#27) in <Unknown source&gt; at line number 
 27 
  at 
 com.script.javascript.RhinoScriptEngine.invoke(RhinoScriptEngine.java:107) 
  at 
 com.script.javascript.RhinoScriptEngine.invokeMethod(RhinoScriptEngine.java:3) 
  at ra.c.d(TtsPluginUiEngine.kt:35) 
  at u9.y0$a$c.invokeSuspend(PluginTtsUI.kt:34) 
  at ff.a.resumeWith(ContinuationImpl.kt:9) 
  at eg.p0.run(DispatchedTask.kt:111) 
  at jg.j$a.run(LimitedDispatcher.kt:4) 
  at lg.j.run(Tasks.kt:3) 
  at lg.a$a.run(CoroutineScheduler.kt:96) 
 Caused by: org.mozilla.javascript.JavaScriptException: FAILED: status=401 
 (<Unknown source&gt;#27) 
  at org.mozilla.javascript.Interpreter.interpretLoop(Interpreter.java:77) 
  at org.mozilla.javascript.Interpreter.interpret(Interpreter.java:57) 
  at 
 org.mozilla.javascript.InterpretedFunction.call(InterpretedFunction.java:21) 
  at org.mozilla.javascript.ContextFactory.doTopCall(ContextFactory.java:1) 
  at 
 com.script.javascript.RhinoScriptEngine$1.superDoTopCall(RhinoScriptEngine.java:1) 
  at 
 com.script.javascript.RhinoScriptEngine$1.doTopCall(RhinoScriptEngine.java:40) 
  at org.mozilla.javascript.ScriptRuntime.doTopCall(ScriptRuntime.java:8) 
  at 
 org.mozilla.javascript.InterpretedFunction.call(InterpretedFunction.java:16) 
  at 
 com.script.javascript.RhinoScriptEngine.invoke(RhinoScriptEngine.java:60) 
  ... 8 more 
 
 
 https://bin.kv2.dev/~65b7b2205de7ed6bfdbca56f 
 
 Dan Frist 
 Founder 
 Sttark.com|864-906-0300 (SMS) 
 2 TASK INDUSTRIAL CT | GREENVILLE SC 29607 
 
 On Mon, Jan 29, 2024, 8:53 AM Jing ***@***.***&gt; wrote: 
 
 &gt; Use devlopment version from Github actions 
 &gt; 
 &gt; — 
 &gt; Reply to this email directly, view it on GitHub 
 &gt; <https://github.com/jing332/tts-server-android/discussions/143#discussioncomment-8283141&gt;, 
 &gt; or unsubscribe 
 &gt; <https://github.com/notifications/unsubscribe-auth/AABZA6S4X7KCWL4BSXLUQHLYQ6SUXAVCNFSM6AAAAABCI65PC6VHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4DEOBTGE2DC&gt; 
 &gt; . 
 &gt; You are receiving this because you authored the thread.Message ID: 
 &gt; ***@***.*** 
 &gt; com&gt; 
 &gt;

**Comment by danfrist** [comment]
![Screenshot_20240129-104730](https://github.com/jing332/tts-server-android/assets/233594/46867a1e-91df-4f2f-be36-c0918dd64f23)

Yes. I also tested the connection with https://reqbin.com/ and found that the voice name needs to be nova not Nova so I made that change in your app, but it still does not work.

<img width="1073" alt="Screenshot 2024-01-29 at 10 53 30 AM" src="https://github.com/jing332/tts-server-android/assets/233594/4eb66359-90be-4373-b690-e88cceb5e5a2">

**Comment by jing332** [comment]
Sorry, the problem is from me,
Please change the first line of the plugin code to 

``` javascript
let apiKey = ttsrv.userVars["apiKey"]
```

**Comment by danfrist** [comment]
Okay. We have made progress. I can now use the audition button and it works properly. 

However, I have selected TTS server as my preferred speech engine in my phone settings. But when I select text to speak, I hear nothing. Eventually I get a system message that TTS server continues to stop. (I think I had this trouble before, but I don't remember how I resolved it then, since it was some time ago.) 

Any more help would be much appreciated.

If you have already made comments in another discussion on how to resolve this I apologize, I cannot read other languages.

**Comment by danfrist** [comment]
I just used the latest stable release and it's working! Thank you.

**Comment by danfrist** [comment]
Unfortunately, the sample rate is not changing for the return from this plugin. Can you help?

**Comment by jing332** [comment]
``` javascript
'getAudioSampleRate': function (locale, voice) {
       return 24000
},
```

The sample rate will be changed the next time you save the TTS configuration.

**Comment by mendelj3** [comment]
Hi, I'm a bit of an amateur and I'd love to get the ChatGPT voice working. If you have time to give me step by step instructions, that would be greatly appreciated. 🙏🏼 

I've added my OpenAI API key, but I get this error when previewing the voice:
https://bin.kv2.dev/~65d2c26d5de7ed6bfdbdcb8b

(Thank you @jing332 for this amazing app! I've been using the Microsoft Neural voice to read out notifications with MacroDroid)

**Reply by danfrist** [reply]
What version of the app do you have? I am using 0.9_202307261647, but unfortunately, I can't find it for download here any longer. It says it's expired. It looks like the only thing available is the [test versions](https://github.com/jing332/tts-server-android/actions/workflows/test.yml). Unfortunately, the last time I tried to use the latest, it did not work. (And I'm not going to try now because I already spent a couple of hours getting it to work like it does right now. 

Anyway, you can gather how to do it from this thread. But you basically need to paste the following code into a new plugin from the plugin manager. Then you need to add your GPT API key by tapping on your plugin in the plugin manager list.

```
let apiKey = ttsrv.userVars["apiKey"]
let speedVar = ttsrv.userVars["speedVar"]

let PluginJS = {
    'name': 'ChatGPT',
    'id': 'speech.openai.com',
    'author': 'TTS Server',
    'description': '',
    'version': 1,
    "vars": {
        apiKey: {label: "API-KEY", hint: "OpenAI API-KEY"},
        speedVar: {label: "Speed", hint: "0.25 to 4.0"},
    },
    'getAudio': function (text, locale, voice, speed, volume, pitch) {
        let reqHeaders = {
            'Content-Type': 'application/json',
            'Authorization': 'Bearer ' + apiKey
        }

        let body = {"model":"tts-1", "input":text, "voice":voice, "speed": speedVar, "response_format":"aac"}
        let str = JSON.stringify(body)
        let resp = ttsrv.httpPost('https://api.openai.com/v1/audio/speech', str, reqHeaders)

        if (resp.isSuccessful()) {
            return resp.body().byteStream()
        } else {
            throw "FAILED: status=" + resp.code()
        }
    }
}

let EditorJS = {
    "getAudioSampleRate": function (locale, voice) {
    
    },

    'onLoadData': function () {

    },

    'getLocales': function () {
        return ['en-US']
    },

    'getVoices': function (locale) {
      return {
        'nova': 'nova',
        'onyx': 'onyx',
        'echo': 'echo'
      }
    },

    'onLoadUI': function (ctx, linearLayout) {

    }

}
```

If you've got specific questions, let me know and I'll try to help.

**Reply by mendelj3** [reply]
Thanks for responding!

My app version is 0.9_202307261647

When I paste that code in and press preview I get an error:
org.mozilla.javascript. JavaScriptException: FAILED: status=429 (<Unknown source>#27) in <Unknown source> at line number 27
(https://bin.kv2.dev/~65d3ee615de7ed6bfdbdd94f)

**Reply by jing332** [reply]
HTTP 429 requests are too many, try using a VPN

**Reply by danfrist** [reply]
What did you enter for the speed, below your GPT API key?
![Screenshot_20240221-082458](https://github.com/jing332/tts-server-android/assets/233594/bdebebea-172a-4e4a-aedf-39b692505377)

Also, try using reqbin as seen above with your variables and see if you can get a 200 response. If not the problem is with your variables or GPT, not this app.

https://github.com/jing332/tts-server-android/discussions/143#discussioncomment-8285567

**Reply by laughingstar101** [reply]
> HTTP 429 requests are too many, try using a VPN

Hey, im getting same error 429, and i already tried using a vpn but it still doesnt work. Pops up error 429 still. Any idea what to do? 
Ps. Thank you so much for developing the app! I use it everyday and it's a lifesaver truly

**Comment by jing332** [comment]
https://github.com/jing332/tts-server-android?tab=readme-ov-file#actions-mirror

**Comment by danfrist** [comment]
I found that if I start the system tts forwarder I don't get any errors

**Reply by Ronald1901** [reply]
I can't get it to work. It gives error 429 every time. What did you do?

**Comment by vivekxter** [comment]
I am also facing this same error 429

com.script.ScriptException: org.mozilla.javascript.JavaScriptException: FAILED: status=429 (<Unknown source>#14) in <Unknown source> at line number 14
 at com.script.javascript.RhinoScriptEngine.invoke(rhinoscriptengine.java:105)
 at com.script.javascript.RhinoScriptEngine.invokeMethod(rhinoscriptengine.java:3)
 at u4.a.b(TtsPluginEngine.kt:89)
 at com.github.jing332.tts_server_android.model.speech.tts.PluginTTS.getAudio(PluginTTS.kt:36)
 at com.github.jing332.tts_server_android.model.speech.tts.e.getAudioWithSystemParams(ITextToSpeechEngine.kt:23)
 at com.github.jing332.tts_server_android.model.speech.tts.e.getAudioWithSystemParams$default(ITextToSpeechEngine.kt:14)
 at t5.c.invokeSuspend(PluginTtsEditViewModel.kt:92)
 at t5.c.invoke(PluginTtsEditViewModel.kt:11)
 at c6.m$b.invokeSuspend(HandlerUtils.kt:29)
 at cc.a.resumeWith(ContinuationImpl.kt:9)
 at [tc.j0.run](https://tc.j0.run/)(DispatchedTask.kt:112)
 at yc.h$[a.run](https://a.run/)(LimitedDispatcher.kt:4)
 at [ad.k.run](https://ad.k.run/)(Tasks.kt:3)
 at ad.a$[a.run](https://a.run/)(CoroutineScheduler.kt:96)
Caused by: org.mozilla.javascript.JavaScriptException: FAILED: status=429 (<Unknown source>#14)
 at org.mozilla.javascript.Interpreter.interpretLoop(interpreter.java:77)
 at org.mozilla.javascript.Interpreter.interpret(interpreter.java:57)
 at [org.mozilla.javascript.InterpretedFunction.call](https://org.mozilla.javascript.interpretedfunction.call/)(interpretedfunction.java:21)
 at org.mozilla.javascript.ContextFactory.doTopCall(contextfactory.java:1)
 at com.script.javascript.RhinoScriptEngine$1.superDoTopCall(rhinoscriptengine.java:1)
 at com.script.javascript.RhinoScriptEngine$1.doTopCall(rhinoscriptengine.java:40)
 at org.mozilla.javascript.ScriptRuntime.doTopCall(scriptruntime.java:8)
 at [org.mozilla.javascript.InterpretedFunction.call](https://org.mozilla.javascript.interpretedfunction.call/)(interpretedfunction.java:16)
 at com.script.javascript.RhinoScriptEngine.invoke(rhinoscriptengine.java:55)
 ... 13 more

**Comment by IT-RT** [comment]
Hey guys, so apparently one of the issue comes from the 429 error not being detailed enough. I'm only I tinkerer with no technical background but I had GPT correct the plug-in for me and got it to work. Please note that you're most likely need to add credit to your API account on open ai. I have added $5 and I was able to test it successfully immediately after. 

```
// --- OpenAI TTS Plugin for TTS Server (jing332) ---
// Changes vs your original:
// 1) Shows OpenAI error body (so 429 becomes actionable)
// 2) Parses speed as a number (0.25–4.0) with safe defaults
// 3) Adds simple exponential backoff retry for HTTP 429
// 4) Expands voice list (keeps your original three + common OpenAI ones)
// 5) Minor robustness: reads userVars inside getAudio so edits apply immediately

let PluginJS = {
  name: "OpenAI TTS (ChatGPT)",
  id: "openai.tts.api",
  author: "TTS Server (modified)",
  description: "OpenAI /v1/audio/speech via API key. Includes retries + better error output.",
  version: 2,

  vars: {
    apiKey: { label: "API-KEY", hint: "OpenAI API key (starts with sk-...)" },
    model: { label: "Model", hint: "e.g. tts-1 or tts-1-hd", default: "tts-1" },
    speed: { label: "Speed", hint: "0.25 to 4.0 (number). Leave blank for 1.0", default: "1.0" },
    format: { label: "Format", hint: "aac / mp3 / opus / wav (try aac first)", default: "aac" },
    retries429: { label: "429 Retries", hint: "How many times to retry on 429 (0–6)", default: "4" }
  },

  getAudio: function (text, locale, voice, speed, volume, pitch) {
    // Always read latest values (so you don’t need to reload plugin after editing vars)
    let apiKey = (ttsrv.userVars["apiKey"] || "").trim();
    let model = (ttsrv.userVars["model"] || "tts-1").trim();
    let format = (ttsrv.userVars["format"] || "aac").trim();
    let speedVarRaw = (ttsrv.userVars["speed"] || "1.0").toString().trim();
    let retries429Raw = (ttsrv.userVars["retries429"] || "4").toString().trim();

    if (!apiKey) {
      throw "FAILED: Missing API key (set Variables → API-KEY)";
    }

    // Parse and clamp speed (OpenAI expects a number 0.25–4.0)
    let speedNum = parseFloat(speedVarRaw);
    if (isNaN(speedNum)) speedNum = 1.0;
    if (speedNum < 0.25) speedNum = 0.25;
    if (speedNum > 4.0) speedNum = 4.0;

    // Retry count for 429
    let maxRetries429 = parseInt(retries429Raw, 10);
    if (isNaN(maxRetries429)) maxRetries429 = 4;
    if (maxRetries429 < 0) maxRetries429 = 0;
    if (maxRetries429 > 6) maxRetries429 = 6;

    let reqHeaders = {
      "Content-Type": "application/json",
      "Authorization": "Bearer " + apiKey
    };

    let bodyObj = {
      model: model,
      input: text,
      voice: voice,
      speed: speedNum,
      response_format: format
    };

    let payload = JSON.stringify(bodyObj);

    // Exponential backoff for 429: 250ms, 500ms, 1000ms, 2000ms...
    // Note: Some TTS Server builds may not expose ttsrv.sleep(). If missing, we retry immediately.
    let attempt = 0;

    while (true) {
      let resp = ttsrv.httpPost("https://api.openai.com/v1/audio/speech", payload, reqHeaders);

      if (resp.isSuccessful()) {
        return resp.body().byteStream();
      }

      let code = resp.code();

      // Try to extract OpenAI’s JSON error body (this is where you’ll see rate_limit vs insufficient_quota)
      let respText = "<no body>";
      try {
        respText = resp.body().string();
      } catch (e) {
        // ignore
      }

      if (code === 429 && attempt < maxRetries429) {
        let waitMs = 250 * Math.pow(2, attempt); // 250, 500, 1000, 2000...
        attempt++;

        // Sleep if the host provides it
        try {
          if (typeof ttsrv.sleep === "function") {
            ttsrv.sleep(waitMs);
          }
        } catch (e2) {
          // ignore
        }

        continue;
      }

      throw "FAILED: status=" + code + " body=" + respText;
    }
  }
};

let EditorJS = {
  getAudioSampleRate: function (locale, voice) {
    // Optional: leave undefined; many engines rely on container sniffing
  },

  onLoadData: function () {},

  getLocales: function () {
    return ["en-US"];
  },

  getVoices: function (locale) {
    // Keep your original three, plus commonly used OpenAI TTS voices
    // (If your account/model doesn’t support a given voice, OpenAI will return an error body.)
    return {
      alloy: "alloy",
      ash: "ash",
      ballad: "ballad",
      coral: "coral",
      echo: "echo",
      fable: "fable",
      nova: "nova",
      onyx: "onyx",
      sage: "sage",
      shimmer: "shimmer"
    };
  },

  onLoadUI: function (ctx, linearLayout) {}
};
```
